src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,168 @@
+package jenkins.plugins.git;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.EnvVars;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * A class which allows {@link AbstractGitSCMSource} to estimate the size of a repository from a distance
+ * without requiring a local checkout.
+ */
+public class GitRepoSizeEstimator {
```

Comment:
The name is still a bit confusing: `GitRepoSizeEstimator` but it only offers the Git Tool to use, nothing about Size

Reply:
I don't mean to change the name, but offer the size data as well

Reply:
My assumption was that the class should only provide the right recommendation for git implementation, to create a more coherent name, I suggest `GitToolAdvisor` or maybe something a bit less dramatic.

Reply:
Do you think we should also return size? While writing the class I couldn't imagine the benefit of providing the size of the repository.

Reply:
it's weird and confusing to have a `GitRepoSizeEstimator` class that is estimating something that is never returned and instead it's returning the tool to use. Something we can change in the last minute

Reply:
I agree, `GitToolSuggestion`? A `GitToolSuggestion.getGitTool()` seems like a good way to tell that the developer that we wish to suggest a better approach in terms of git implementation.

Reply:
I'm really bad at naming methods and classes, so I will let others suggest something. for me it's fine if the class refers an actual behaviour